### PR TITLE
feat: tee port for scheduled API

### PR DESCRIPTION
**Pull request by discord9**
- Add `tee()` method for `RecvPort<TeeingHandoff<T>>` so it can be cloned accordingly, I'm not sure if I made prec and succ fields right though.

- Added `add_handoff` and `get_handoff_by_id` for Hydroflow for a more detailed control of handoffs

- also added test `test_auto_tee` back. So `tee` method is tested.

### CODE REVIEW

**Comment by discord9** [comment]
> Will get to playing around with this eventually, feel free to ping me again

The thing is, the scheduled API needs a simple way to copy recv port, similar to how you can do it with differential dataflow's collection. Also, for a specific situation where someone can't use the entire Rust compiler and has to use the scheduled API instead, how would you create the tee API for scheduled api?

**Comment by MingweiSamuel** [comment]
So without the teeing handoff the way you'd make a tee is manually by adding a one-in two-out subgraph that clones the value and writes to both outputs. Example in test.rs:

https://github.com/hydro-project/hydroflow/blob/b364328dd3a37cf788b7197e1e5cbc7e10cce43d/hydroflow/tests/test.rs#L219-L230

(Can also do this as part of a bigger subgraph of course, for better inlining)

**Comment by discord9** [comment]
> So without the teeing handoff the way you'd make a tee is manually by adding a one-in two-out subgraph that clones the value and writes to both outputs. Example in test.rs:
> 
> https://github.com/hydro-project/hydroflow/blob/b364328dd3a37cf788b7197e1e5cbc7e10cce43d/hydroflow/tests/test.rs#L219-L230
> 
> (Can also do this as part of a bigger subgraph of course, for better inlining)

Indeed, I try this method in the system I try to build, and found out it can be a headacahe to manage it if my graph need to be more dynamic or need to build from top-down(hence need to know how many downstream port is being send to, which is kind a messy), so a teeing handoff could be very useful here

**Comment by MingweiSamuel** [comment]
Test that fails:
https://github.com/MingweiSamuel/hydroflow/commit/84968ea9d07550138c0f5850ec9d75b9ea18efb8#diff-291410ca616992f7d6090fd9dd65dc2e7aa1f3c6b7a80ce5f33d4fe43f0bc5caR93-R104

**Comment by discord9** [comment]
> Have looked and tested this a bit, PR implementation has a bug that means it will forget to schedule child subgraphs (after the initial run). Tricky bit for implementing this is that current code assumes both ends of the handoffs are the same handoff ID, and teeing has each piece with a different ID. So the mechanism for building the preds/succs misses those implicit links. Need to either find a way to treat teeing as a single handoff ID (and somehow differentiate between the consumers) or make the preds/succs mechanism more general to handle those links

Sorry for the late reply, something in real life consume all my time, anyway, I think I fixed the bug by "find a way to treat teeing as a single handoff ID":

I basically found all child `handoff`s that are being `tee`d from root handoff, and check their succ subgraph when trying to determine which subgraph to run.
, I also impl `TeeHandoff`'s `is_bottom` method so it actually will schedule downstream subgraph(before the fix it always think it's empty) , this should fixs the bug now?

**Comment by discord9** [comment]
@MingweiSamuel Please Take a look?

**Comment by MingweiSamuel** [comment]
Thank you for your patience, taking a look, I may modify it a bit but should be pretty close to merging

**Comment by discord9** [comment]
> Thank you for your patience, taking a look, I may modify it a bit but should be pretty close to merging

I realize in some case after some teeing recv port, the last recv port might need to be manually released so it wouldn't overflow it's buffer without anyone receiving it, so I came up with a solution that give `ReaderHandoff` a alive flag, feel free to change it. Also I rebase the branch on hydroflow 6.0.0 now

File: hydroflow/src/scheduled/graph.rs [hunk 1]
```diff
@@ -216,11 +258,18 @@ impl<'a> Hydroflow<'a> {
...
-                    for &succ_id in handoff.succs.iter() {
+                    // find out all child handoff spawn from teeing(if any)
+                    let mut all_succs = vec![handoff.succs.iter()];
+                    if let Some(childs) = self.handoff_tee_childs.get(&handoff_id) {
+                        for ch in childs.iter() {
+                            all_succs.push(self.handoffs[ch.0].succs.iter());
+                        }
+                    }
+
+                    for &succ_id in all_succs.iter().flat_map(|it| it.clone()) {
```
Comment (MingweiSamuel) [inline hydroflow/src/scheduled/graph.rs:272]: (should be `all_succs.into_iter().flatten()` but will be refactoring this away)

**Comment by discord9** [comment]
would it be wise to call panic in teeing recv port's drop? Since a teeing recv port that wouldn't be use will surely cause memory leaking?

**Comment by MingweiSamuel** [comment]
Yeah, do you think you'll actually be teeing then not using the new output? Should be an error or warning to do that

**Comment by discord9** [comment]
> Yeah, do you think you'll actually be teeing then not using the new output? Should be an error or warning to do that

More like I wouldn't be using the root handoff, the usecase could look like this:
a map with key being id and value being teeing handoff, and during building a scheduled dataflow, handoff are being teeing repeatly, after the graph is built, the `map<key, teeing handoff>` is then dropped, so a panic could make sense because the error would be in schedule phase, not during evaluation so it could be easier to found out by testing. But a simple warning could also do the job, since such misuse is supposed to be detected in unit test or so, and some may prefer not to panic the entire application. So a simple warning could be good enough.

File: hydroflow/src/scheduled/graph.rs [hunk 2]
```diff
@@ -80,6 +81,73 @@ impl<'a> Default for Hydroflow<'a> {
...
+            // Set up teeing metadata.
+            // If we're teeing from a child make sure to find root.
+            let tee_root = self.handoffs[tee_parent_port.handoff_id.0].pred_handoffs[0];
+            // Go to `tee_root`'s successors and insert self.
+            self.handoffs[tee_root.0].succ_handoffs.push(new_handoff_id);
+            // Set self's predecessor as `tee_root`.
+            self.handoffs[new_handoff_id.0].pred_handoffs = vec![tee_root];
+
+            new_handoff_id
+        };
```
Comment (MingweiSamuel) [inline hydroflow/src/scheduled/graph.rs:124]: cleanup bit of redundant code in this path

File: hydroflow/src/scheduled/graph.rs [hunk 3]
```diff
@@ -216,11 +284,18 @@ impl<'a> Hydroflow<'a> {
...
-                    for &succ_id in handoff.succs.iter() {
+                    // TODO(mingwei): WIP refactor this away
+                    // find out all child handoff spawn from teeing(if any)
+                    let mut all_succs = vec![handoff.succs.iter()];
+                    for succ_handoff in self.handoffs[handoff_id.0].succ_handoffs.iter() {
+                        all_succs.push(self.handoffs[succ_handoff.0].succs.iter());
+                    }
+
+                    for &succ_id in all_succs.into_iter().flatten() {
+                        // handoff.succs.iter() {
```
Comment (MingweiSamuel) [inline hydroflow/src/scheduled/graph.rs:298]: should be not needed now, hopefully....

File: hydroflow/src/scheduled/graph.rs [hunk 4]
```diff
@@ -725,8 +813,23 @@ pub struct HandoffData {
     pub(super) name: Cow<'static, str>,
     /// Crate-visible to crate for `handoff_list` internals.
     pub(super) handoff: Box<dyn HandoffMeta>,
+    /// Preceeding subgraphs (including the send side of a teeing handoff).
     pub(super) preds: SmallVec<[SubgraphId; 1]>,
+    /// Successor subgraphs (including recv sides of teeing handoffs).
     pub(super) succs: SmallVec<[SubgraphId; 1]>,
+
+    /// Predacessor handoffs, used by teeing handoffs.
```
Comment (MingweiSamuel) [inline hydroflow/src/scheduled/graph.rs:821]: Predecessor

**Review by MingweiSamuel — CHANGES_REQUESTED** [review]
notes to self

File: hydroflow/src/scheduled/graph.rs (hunk 3 above)
Comment (MingweiSamuel) [reply hydroflow/src/scheduled/graph.rs:298]: Yup :)

**Review by MingweiSamuel — APPROVED** [review]

**Comment by MingweiSamuel** [comment]
Doesn't panic or really warn if you forget to use a tee output though, to note

**Pull request merged**